File: clients/eucli-studio/src/domain/release.ts
import { normalizeLocalBoxState, type LocalBoxState } from './localBox'

export type EucliBoxCompatibility = {
  minimumVersion: string
  maximumVersionExclusive: string
}

export type CompatibilityStatus = {
  compatible: boolean
  reason: string
  currentEucliBoxVersion: string
  requiredEucliBoxCompatibility: EucliBoxCompatibility
}

export type ReleaseArtifactIdentity = {
  kind: 'eucli-box' | 'tool' | 'plugin' | string
  id: string
}

export type OfficialReleaseSource = {
  kind: string
  repository: string
  owner: string
  name: string
}

export type ReleaseCheckResult = {
  artifact: ReleaseArtifactIdentity
  source: OfficialReleaseSource
  installed: boolean
  currentVersion: string
  latestVersion: string
  status: 'not_checked' | 'checking' | 'completed' | 'failed' | string
  checkedAt: string
  updateAvailable: boolean
  releaseUrl: string
  releaseNotes: string
  downloadSize: number
  compatibility: CompatibilityStatus | null
  affectedArtifacts: ReleaseArtifactIdentity[]
  failureReason: string
}

export type ReleaseCheckSnapshot = {
  status: 'not_checked' | 'checking' | 'completed' | 'failed' | string
  source?: string
  startedAt: string
  checkedAt: string
  results: ReleaseCheckResult[]
  failureReason: string
}

export type ReleaseOperationProgress = {
  receivedBytes: number
  totalBytes: number
}

export type ReleaseOperationError = {
  code: string
  phase: string
  message: string
}

export type ArtifactInstallState = {
  operationId: string
  artifact: ReleaseArtifactIdentity
  installed: boolean
  currentVersion: string
  targetVersion: string
  status: string
  phase: string
  progress: ReleaseOperationProgress
  error: ReleaseOperationError
}

export type ArtifactActivityState = {
  artifact: ReleaseArtifactIdentity
  active: boolean
  activeRequests: number
  updating: boolean
  reason: string
}

export const artifactStatusLabels: Record<string, string> = {
  not_installed: '未安装',
  checking_release: '检查发行中',
  ready_to_install: '可安装',
  ready_to_update: '可更新',
  downloading: '下载中',
  verifying: '核对中',
  checking_activity: '检查活动中',
  preparing: '准备中',
  switching: '切换中',
  starting: '启动中',
  active: '已启用',
  unavailable: '不适用',
  failed: '失败',
  blocked: '被阻止',
  restoring: '恢复中',
}

export function normalizeArtifactInstallState(value: unknown): ArtifactInstallState {
  const source = objectValue(value)
  return {
    operationId: text(source.operationId),
    artifact: normalizeReleaseArtifactIdentity(source.artifact),
    installed: source.installed === true,
    currentVersion: text(source.currentVersion),
    targetVersion: text(source.targetVersion),
    status: text(source.status),
    phase: text(source.phase),
    progress: {
      receivedBytes: finiteNumber((source as any).progress?.receivedBytes),
      totalBytes: finiteNumber((source as any).progress?.totalBytes),
    },
    error: {
      code: text((source as any).error?.code),
      phase: text((source as any).error?.phase),
      message: text((source as any).error?.message),
    },
  }
}

export function normalizeArtifactActivityState(value: unknown): ArtifactActivityState {
  const source = objectValue(value)
  return {
    artifact: normalizeReleaseArtifactIdentity(source.artifact),
    active: source.active === true,
    activeRequests: finiteNumber(source.activeRequests),
    updating: source.updating === true,
    reason: text(source.reason),
  }
}

export type StudioBootstrap = {
  clientVersion: string
  clientEucliBoxCompatibility: EucliBoxCompatibility
  localBox: LocalBoxState
  eucliBoxConfigured: boolean
  eucliBoxReachable: boolean
  eucliBoxUrl: string
  eucliBoxVersion: string
  eucliBoxCompatibility: CompatibilityStatus | null
  businessAvailable: boolean
  eucliBoxIssue: string
  releaseChecks: ReleaseCheckSnapshot
}

export function normalizeEucliBoxCompatibility(value: unknown): EucliBoxCompatibility {
  const source = objectValue(value)
  return {
    minimumVersion: text(source.minimumVersion),
    maximumVersionExclusive: text(source.maximumVersionExclusive),
  }
}

export function normalizeCompatibilityStatus(value: unknown): CompatibilityStatus {
  const source = objectValue(value)
  return {
    compatible: source.compatible === true,
    reason: text(source.reason),
    currentEucliBoxVersion: text(source.currentEucliBoxVersion),
    requiredEucliBoxCompatibility: normalizeEucliBoxCompatibility(source.requiredEucliBoxCompatibility),
  }
}

export function normalizeStudioBootstrap(value: unknown): StudioBootstrap {
  const source = objectValue(value)
  return {
    clientVersion: text(source.clientVersion),
    clientEucliBoxCompatibility: normalizeEucliBoxCompatibility(source.clientEucliBoxCompatibility),
    localBox: normalizeLocalBoxState(source.localBox),
    eucliBoxConfigured: source.eucliBoxConfigured === true,
    eucliBoxReachable: source.eucliBoxReachable === true,
    eucliBoxUrl: text(source.eucliBoxUrl),
    eucliBoxVersion: text(source.eucliBoxVersion),
    eucliBoxCompatibility: source.eucliBoxCompatibility && typeof source.eucliBoxCompatibility === 'object'
      ? normalizeCompatibilityStatus(source.eucliBoxCompatibility)
      : null,
    businessAvailable: source.businessAvailable === true,
    eucliBoxIssue: text(source.eucliBoxIssue),
    releaseChecks: normalizeReleaseCheckSnapshot(source.releaseChecks),
  }
}

export function normalizeReleaseCheckSnapshot(value: unknown): ReleaseCheckSnapshot {
  const source = objectValue(value)
  return {
    status: text(source.status) || 'not_checked',
    source: text(source.source),
    startedAt: text(source.startedAt),
    checkedAt: text(source.checkedAt),
    results: Array.isArray(source.results) ? source.results.map(normalizeReleaseCheckResult) : [],
    failureReason: text(source.failureReason),
  }
}

function normalizeReleaseCheckResult(value: unknown): ReleaseCheckResult {
  const source = objectValue(value)
  return {
    artifact: normalizeReleaseArtifactIdentity(source.artifact),
    source: normalizeOfficialReleaseSource(source.source),
    installed: source.installed === true,
    currentVersion: text(source.currentVersion),
    latestVersion: text(source.latestVersion),
    status: text(source.status) || 'not_checked',
    checkedAt: text(source.checkedAt),
    updateAvailable: source.updateAvailable === true,
    releaseUrl: text(source.releaseUrl),
    releaseNotes: text(source.releaseNotes),
    downloadSize: finiteNumber(source.downloadSize),
    compatibility: source.compatibility && typeof source.compatibility === 'object'
      ? normalizeCompatibilityStatus(source.compatibility)
      : null,
    affectedArtifacts: Array.isArray(source.affectedArtifacts)
      ? source.affectedArtifacts.map(normalizeReleaseArtifactIdentity)
      : [],
    failureReason: text(source.failureReason),
  }
}

function normalizeReleaseArtifactIdentity(value: unknown): ReleaseArtifactIdentity {
  const source = objectValue(value)
  return { kind: text(source.kind), id: text(source.id) }
}

function normalizeOfficialReleaseSource(value: unknown): OfficialReleaseSource {
  const source = objectValue(value)
  return {
    kind: text(source.kind),
    repository: text(source.repository),
    owner: text(source.owner),
    name: text(source.name),
  }
}

export function compatibilityRangeText(value: EucliBoxCompatibility | null | undefined): string {
  const minimum = text(value?.minimumVersion)
  const maximum = text(value?.maximumVersionExclusive)
  if (!minimum || !maximum) return '范围资料无效'
  return `[${minimum}, ${maximum})`
}

function objectValue(value: unknown): Record<string, any> {
  return value && typeof value === 'object' && !Array.isArray(value) ? value as Record<string, any> : {}
}

function text(value: unknown): string {
  return String(value ?? '').trim()
}

function finiteNumber(value: unknown): number {
  const number = Number(value)
  return Number.isFinite(number) && number >= 0 ? number : 0
}

File: pkg/types/release.go
package types

import "time"

type EucliBoxCompatibility struct {
	MinimumVersion          string `json:"minimumVersion"`
	MaximumVersionExclusive string `json:"maximumVersionExclusive"`
}

type CompatibilityStatus struct {
	Compatible                    bool                  `json:"compatible"`
	Reason                        string                `json:"reason,omitempty"`
	CurrentEucliBoxVersion        string                `json:"currentEucliBoxVersion,omitempty"`
	RequiredEucliBoxCompatibility EucliBoxCompatibility `json:"requiredEucliBoxCompatibility"`
}

type EucliBoxRelease struct {
	Version     string `json:"version"`
	DataVersion string `json:"dataVersion"`
}

type EucliBoxReleaseInfo struct {
	Version             string               `json:"version"`
	DataVersion         string               `json:"dataVersion"`
	ClientCompatibility *CompatibilityStatus `json:"clientCompatibility,omitempty"`
}

const (
	ReleaseArtifactKindBox    = "eucli-box"
	ReleaseArtifactKindTool   = "tool"
	ReleaseArtifactKindPlugin = "plugin"

	ReleasePlatformWindowsX64 = "windows-x64"

	ReleaseCandidateStatusCompleted = "completed"
	ReleaseCandidateStatusFailed    = "failed"
)

type ReleaseArtifactIdentity struct {
	Kind string `json:"kind"`
	ID   string `json:"id"`
}

type OfficialReleaseSource struct {
	Kind       string `json:"kind"`
	Repository string `json:"repository"`
	Owner      string `json:"owner"`
	Name       string `json:"name"`
	Ref        string `json:"ref,omitempty"`
}

type ReleaseSourceRecord struct {
	Repository string `json:"repository"`
	Commit     string `json:"commit"`
	Recorded   bool   `json:"recorded"`
}

type ReleaseFileRecord struct {
	Name   string `json:"name"`
	Size   int64  `json:"size"`
	SHA256 string `json:"sha256"`
}

type ReleaseExternalAsset struct {
	Name        string `json:"name"`
	Source      string `json:"source"`
	Version     string `json:"version"`
	PackagePath string `json:"packagePath"`
	FileCount   int    `json:"fileCount"`
	TreeSHA256  string `json:"treeSha256"`
}

type ReleaseProductRecord struct {
	SchemaVersion    int                     `json:"schemaVersion"`
	Artifact         ReleaseArtifactIdentity `json:"artifact"`
	Version          string                  `json:"version"`
	Platform         string                  `json:"platform"`
	OfficialSource   string                  `json:"officialSource"`
	Compatibility    *EucliBoxCompatibility  `json:"eucliBoxCompatibility,omitempty"`
	Source           ReleaseSourceRecord     `json:"source"`
	DataVersion      string                  `json:"dataVersion,omitempty"`
	ExternalAssets   []ReleaseExternalAsset  `json:"externalAssets,omitempty"`
}

type ReleaseManifest struct {
	SchemaVersion    int                     `json:"schemaVersion"`
	Artifact         ReleaseArtifactIdentity `json:"artifact"`
	Version          string                  `json:"version"`
	Platform         string                  `json:"platform"`
	TagName          string                  `json:"tagName"`
	OfficialSource   string                  `json:"officialSource"`
	Compatibility    *EucliBoxCompatibility  `json:"eucliBoxCompatibility,omitempty"`
	Source           ReleaseSourceRecord     `json:"source"`
	DataVersion      string                  `json:"dataVersion,omitempty"`
	ExternalAssets   []ReleaseExternalAsset  `json:"externalAssets,omitempty"`
	Archive          ReleaseFileRecord       `json:"archive"`
	Files            []ReleaseFileRecord     `json:"files"`
}

// ArtifactInstallation 是业务端当前真实的已装事实：某个发布物装了、装的是哪个版本、适用什么范围。
// FailureReason 只在读取该分类已装事实失败时出现，此时身份是分类本身。
type ArtifactInstallation struct {
	Artifact      ReleaseArtifactIdentity `json:"artifact"`
	Version       string                  `json:"version"`
	Compatibility *EucliBoxCompatibility  `json:"eucliBoxCompatibility,omitempty"`
	FailureReason string                  `json:"failureReason,omitempty"`
}

// ArtifactInstallationList 是已装事实的完整清单。
type ArtifactInstallationList struct {
	Artifacts []ArtifactInstallation `json:"artifacts"`
}

// ArtifactReleaseCandidate 是某个发布物在某个来源下的可用候选事实，
// 以及它与已装事实比对后的结论；不携带任何快照状态。
type ArtifactReleaseCandidate struct {
	Artifact          ReleaseArtifactIdentity   `json:"artifact"`
	Source            OfficialReleaseSource     `json:"source"`
	Installed         bool                      `json:"installed"`
	CurrentVersion    string                    `json:"currentVersion,omitempty"`
	LatestVersion     string                    `json:"latestVersion,omitempty"`
	Status            string                    `json:"status"`
	PublishedAt       time.Time                 `json:"publishedAt,omitempty"`
	UpdateAvailable   bool                      `json:"updateAvailable"`
	ReleaseURL        string                    `json:"releaseUrl,omitempty"`
	ReleaseNotes      string                    `json:"releaseNotes,omitempty"`
	DownloadSize      int64                     `json:"downloadSize,omitempty"`
	Compatibility     *CompatibilityStatus      `json:"compatibility,omitempty"`
	AffectedArtifacts []ReleaseArtifactIdentity `json:"affectedArtifacts,omitempty"`
	FailureReason     string                    `json:"failureReason,omitempty"`
}

// ArtifactCandidateList 是某个分类在某个来源下的候选完整清单：
// 来源种类（official/local）+ 该分类全部可展示项。
type ArtifactCandidateList struct {
	SourceKind string                    `json:"source"`
	Candidates []ArtifactReleaseCandidate `json:"candidates"`
}

// ArtifactInstallState 表示单个发布物（工具或插件）当前对外展示的整体安装/更新状态。
// Status 与 Phase 是两个互不替代的事实，各自只使用自己的词表，不得混用。
type ArtifactInstallState struct {
	OperationID    string                   `json:"operationId,omitempty"`
	Artifact       ReleaseArtifactIdentity  `json:"artifact"`
	Installed      bool                     `json:"installed"`
	CurrentVersion string                   `json:"currentVersion,omitempty"`
	TargetVersion  string                   `json:"targetVersion,omitempty"`
	Status         string                   `json:"status"`
	Phase          string                   `json:"phase,omitempty"`
	Progress       ReleaseOperationProgress `json:"progress,omitempty"`
	Error          ReleaseOperationError    `json:"error,omitempty"`
}

// ReleaseOperationProgress 是一次安装/更新操作的下载进度事实；
// 总量未知时 TotalBytes 必须为 0，不能填假值。
type ReleaseOperationProgress struct {
	ReceivedBytes int64 `json:"receivedBytes"`
	TotalBytes    int64 `json:"totalBytes"`
}

// ArtifactOperationList 是所有有操作事实的发布物状态清单：
// 运行中的任务与落盘的失败/取消终态记录；无任务的发布物不出现。
type ArtifactOperationList struct {
	Operations []ArtifactInstallState `json:"operations"`
}

type ReleaseOperationError struct {
	Code    string `json:"code,omitempty"`
	Phase   string `json:"phase,omitempty"`
	Message string `json:"message,omitempty"`
}

// ArtifactActivityState 表示发布物当前的真实活动状态，用于展示阻止原因。
type ArtifactActivityState struct {
	Artifact       ReleaseArtifactIdentity `json:"artifact"`
	Active         bool                    `json:"active"`
	ActiveRequests int                     `json:"activeRequests"`
	Updating       bool                    `json:"updating"`
	Reason         string                  `json:"reason,omitempty"`
}

const (
	ArtifactStatusNotInstalled     = "not_installed"
	ArtifactStatusCheckingRelease  = "checking_release"
	ArtifactStatusReadyToInstall   = "ready_to_install"
	ArtifactStatusReadyToUpdate    = "ready_to_update"
	ArtifactStatusDownloading      = "downloading"
	ArtifactStatusVerifying        = "verifying"
	ArtifactStatusCheckingActivity = "checking_activity"
	ArtifactStatusPreparing        = "preparing"
	ArtifactStatusSwitching        = "switching"
	ArtifactStatusStarting         = "starting"
	ArtifactStatusActive           = "active"
	ArtifactStatusUnavailable      = "unavailable"
	ArtifactStatusFailed           = "failed"
	ArtifactStatusBlocked          = "blocked"
	ArtifactStatusCancelled        = "cancelled"
	ArtifactStatusRestoring        = "restoring"
)

const (
	ArtifactPhaseCandidate    = "candidate"
	ArtifactPhaseCompatibility = "compatibility"
	ArtifactPhaseActivity     = "activity"
	ArtifactPhaseDownload     = "download"
	ArtifactPhaseManifest     = "manifest"
	ArtifactPhaseArchive      = "archive"
	ArtifactPhasePackage      = "package"
	ArtifactPhasePrepare      = "prepare"
	ArtifactPhaseSwitch       = "switch"
	ArtifactPhaseProbe        = "probe"
	ArtifactPhaseRestore      = "restore"
	ArtifactPhaseRefresh      = "refresh"
)

// 固定错误码：工具和插件后台、网关、客户端统一使用，不得各自造同义名称。
const (
	ArtifactErrorNotInstalled       = "ARTIFACT_NOT_INSTALLED"
	ArtifactErrorReleaseUnavailable = "ARTIFACT_RELEASE_UNAVAILABLE"
	ArtifactErrorCandidateMismatch  = "ARTIFACT_CANDIDATE_MISMATCH"
	ArtifactErrorCompatibility      = "ARTIFACT_COMPATIBILITY_FAILED"
	ArtifactErrorDownloadFailed     = "ARTIFACT_DOWNLOAD_FAILED"
	ArtifactErrorManifestInvalid    = "ARTIFACT_MANIFEST_INVALID"
	ArtifactErrorPackageInvalid     = "ARTIFACT_PACKAGE_INVALID"
	ArtifactErrorPathInvalid        = "ARTIFACT_PATH_INVALID"
	ArtifactErrorToolActive         = "TOOL_ACTIVE"
	ArtifactErrorPluginActive       = "PLUGIN_ACTIVE"
	ArtifactErrorUpdateInProgress   = "ARTIFACT_UPDATE_IN_PROGRESS"
	ArtifactErrorPrepareFailed      = "ARTIFACT_PREPARE_FAILED"
	ArtifactErrorSwitchFailed       = "ARTIFACT_SWITCH_FAILED"
	ArtifactErrorProbeFailed        = "ARTIFACT_PROBE_FAILED"
	ArtifactErrorRestoreFailed      = "ARTIFACT_RESTORE_FAILED"
	ArtifactErrorStateUnknown       = "ARTIFACT_STATE_UNKNOWN"
	ArtifactErrorDataChanged        = "ARTIFACT_DATA_CHANGED"
	ArtifactErrorCancelRejected     = "ARTIFACT_CANCEL_REJECTED"
)
